### ISSUE TITLE
Add an actual-model AMP forward smoke test for custom architectures.

### ISSUE THREAD

**Issue by Anawaert** [ultralytics/ultralytics#24719]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi Ultralytics,

Recently, I have been training a YOLO26n model with ultralytics and experimenting with some custom architectural modules. In one of my custom modules, I use adaptive average pooling. Since the backward pass of this PyTorch operator on CUDA can be non-deterministic in some cases, I moved this operation to the CPU inside the module. The relevant part of the module's forward() method is implemented like this:

`avg_pool_h_out = self.avg_pool_h(x.cpu()).to(x.device)`

This avoids the deterministic-algorithm issue caused by running the adaptive average pooling backward pass on CUDA. However, I noticed that although `ultralytics` passes the AMP Check at the beginning of training, autocast can still produce FP16 tensors during the actual model forward pass. As a result, FP16 tensors may be moved to the CPU by the custom module, which can lead to PyTorch runtime errors because CPU-side FP16 computation is not fully supported for all operators on typical x86/amd64 CPUs.

My understanding is that the current AMP Check mainly verifies AMP behaviour using a reference YOLO model, rather than running the actual model that is going to be trained. Therefore, it may not detect custom layers whose `forward()` method contains CPU offloading paths, such as `x.cpu()` or `x.to("cpu")`.

Would it be possible to add an additional AMP check for the actual model being trained?

One possible approach I can think of is to keep all existing functions and interfaces unchanged, and add a lightweight method such as:

`check_amp_actual_model_forward(model)`

This function would receive the real model used for training, which can be obtained directly from the training pipeline after constructing it with something like:

`YOLO("actual-model-config-or-weight.format")`

Then it could run a small forward-only smoke test under CUDA autocast, using either a dummy input tensor or the existing `assets/bus.jpg`, and check whether the model forward pass raises any exception or produces non-finite outputs. If an AMP-unsafe custom layer is detected, `ultralytics` could warn the user and set `amp=False` before training starts.

This could help catch issues such as:
- FP16 tensors being moved to CPU inside a custom `forward()` method;
- unsupported CPU-side FP16 operations;
- dtype/device mismatches under AMP;
- non-finite outputs caused by custom AMP-unsafe layers.

### Use case

The goal of this feature would not be to replace the current AMP Check for official Ultralytics models. Official YOLO models are expected to be AMP-safe already. Instead, this would mainly benefit users who customise `ultralytics` architectures by adding new modules. At a relatively low cost, it could help verify whether the actual model about to be trained can at least complete one forward pass under CUDA autocast.

### Additional

Of course, this would only be a forward smoke test rather than a full training-step or backward-pass AMP validation. Still, I think it would be useful for detecting many practical custom-module AMP issues earlier, before the real training loop fails.

Would such an enhancement be acceptable for `ultralytics`? If so, I would be happy to try implementing it and submit a PR later.

Thanks!

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Anawaert, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this is a feature request 💡, thank you for clearly describing the use case, proposed approach, and your willingness to contribute a PR. If you proceed, please include as much implementation detail as possible, along with any relevant logs, test results, and a minimal example showing the current behavior and the proposed validation path under AMP.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by everglow01** [comment]
Real gap here, the thing is `check_amp` validates a clean reference `yolo26n.pt` rather than the model you're actually training, so for anyone with custom modules it passing gives false confidence and the failure only shows up once training starts, which makes a forward smoke test on the real model a sensible addition. A few suggestions though. First, scope it honestly: a forward only test won't catch backward or grad scaler issues like overflow and custom losses, and that's where a lot of AMP NaNs actually come from, so it's worth saying so in the docstring rather than trading one false sense of safety for another. Second, run it in `train()` mode with the real autocast context and a representative imgsz, since your `.cpu()` branch and things like BN/Dropout behave differently when `model.training` is True; just remember to restore the original mode and grad state afterward and wrap it in `torch.no_grad()`. Third, check `torch.isfinite(out).all()` too, not just exceptions, so you catch the silent NaN/Inf cases. And on the failure path I'd lean toward a loud, named warning over silently setting `amp=False`, because the `.cpu()` issue is really a bug in the custom module and silently disabling AMP both hides it and drops the speedup; something like "forward failed under autocast, common cause: FP16 tensors moved to CPU or AMP unsafe ops, AMP disabled for now" points people at the actual fix. One optional idea: once the smoke test passes you could run the existing FP32 vs AMP `allclose` against the real model instead of the reference, which folds precision and coverage into one thing. Happy to review the PR.

**Comment by glenn-jocher** [comment]
This seems like a reasonable enhancement—[`check_amp()`](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp) currently verifies AMP with a reference YOLO26 model, so a minimal forward-only smoke test on the actual training model would help catch custom-module cases like FP16 tensors being moved to CPU; a PR is welcome if you keep it narrowly scoped, restore the original train/eval state after the check, document clearly that it does not validate backward or `GradScaler` behavior, and include a focused test.

**Comment by Anawaert** [comment]
Okay, thanks for your suggestion. I will try to implement it in the `check_amp()` context, and offer a PR with detailed descriptions.

**Comment by glenn-jocher** [comment]
Sounds good — please open the PR when ready with the implementation, a focused regression test for an AMP-unsafe custom layer, and brief notes clarifying that it is a forward smoke test only.
